src/marinade-state/marinade-state.ts

```diff
@@ -133,6 +139,74 @@ export class MarinadeState {
     )
   }
 
+  async getStakeStates(): Promise<StakeState[]> {
+    const stakeAccountInfos = await this.anchorProvider.connection.getProgramAccounts(STAKE_PROGRAM_ID, {
+      filters: [
+        { dataSize: 200 },
+        {
+          memcmp: {
+            offset: 44,
+            bytes: this.marinade.config.stakeWithdrawAuthPDA.toString(),
+          },
+        },
+      ],
+    })
+
+    return stakeAccountInfos.map((stakeAccountInfo) => {
+      const { data } = stakeAccountInfo.account
+      // The data's first 4 bytes are: u8 0x0 0x0 0x0 but borsh uses only the first byte to find the enum's value index.
+      // The next 3 bytes are unused and we need to get rid of them (or somehow fix the BORSH schema?)
+      const adjustedData = Buffer.concat([
+        data.slice(0, 1), // the first byte indexing the enum
+        data.slice(4, data.length), // the first byte indexing the enum
+      ])
+      return deserializeUnchecked(
+        MARINADE_BORSH_SCHEMA,
+        StakeState,
+        adjustedData,
+      )
+    })
+  }
+
+  async getStakeInfos(): Promise<StakeInfo[]> {
+    const stakeRecords = await this.getStakeRecords()
+    const stakeInfos = new Array<StakeInfo>()
+
+    const to_process = stakeRecords.length
+    let processed = 0
+    // rpc.get_multiple_accounts() has a max of 100 accounts
+    const BATCH_SIZE = 100
+    while (processed < to_process) {
+
+      const accountInfos: AccountInfo<Buffer>[] = await this.anchorProvider.connection.getMultipleAccountsInfo(
+        stakeRecords
+          .slice(processed, processed + BATCH_SIZE)
+          .map(stakeRecord => stakeRecord.stakeAccount)
+      ) as AccountInfo<Buffer>[]
+
+      stakeInfos.push(...accountInfos.map((accountInfo, index) => {
+
+        const adjustedData = Buffer.concat([
+          accountInfo?.data.slice(0, 1), // the first byte indexing the enum
+          accountInfo?.data.slice(4, accountInfo?.data.length), // the first byte indexing the enum
+        ])
+
+        return {
+          index: processed + index,
+          record: stakeRecords[processed + index],
+          stake: deserializeUnchecked(
+            MARINADE_BORSH_SCHEMA,
+            StakeState,
+            adjustedData,
+          ),
+          balance: new BN(accountInfo.lamports),
+        } as StakeInfo
```

Comment:
Let's avoid casting when we can.
```suggestion
        return new StakeInfo({
          index: processed + index,
          record: stakeRecords[processed + index],
          stake: deserializeUnchecked(
            MARINADE_BORSH_SCHEMA,
            StakeState,
            adjustedData,
          ),
          balance: new BN(accountInfo.lamports),
        })
```